nupic/research/frameworks/dendrites/dendritic_layers.py

```diff
@@ -101,13 +102,111 @@ class AbsoluteMaxGatingDendriticLayer(BiasingDendriticLayer):
     """
 
     def apply_dendrites(self, y, dendrite_activations):
-        output_max = dendrite_activations.max(dim=2).values
-        output_min = dendrite_activations.min(dim=2).values
-        sign_mask = torch.where(
-            (output_max ** 2) > (output_min ** 2),
-            torch.ones(output_max.shape).to(y.device),
-            -1.0 * torch.ones(output_min.shape).to(y.device)
-        )
-        return y * torch.sigmoid(
-            torch.abs(dendrite_activations).max(dim=2).values * sign_mask
+        inds = dendrite_activations.abs().max(dim=2).indices
+        inds = inds.unsqueeze(dim=2)
+        dendrite_activations = torch.gather(dendrite_activations, dim=2, index=inds)
+        dendrite_activations = dendrite_activations.squeeze()
+        dendrite_activations = torch.sigmoid(dendrite_activations)
+        return y * dendrite_activations
+
+
+class GatingDendriticLayer2d(SparseWeights2d):
+    """
+    A convolutional version of `GatingDendriticLayer`. The multiplicative dendrite
+    outputs are applied element-wise to each output channel. That is, for a given
+    output channel, all activation values (determined by the convolution operation) are
+    multiplied by a single value computed via dendrites.
+    """
+
+    def __init__(
+        self, module, num_segments, dim_context,
+        module_sparsity, dendrite_sparsity, dendrite_bias=None
+    ):
+        """
+        :param module: conv2d module which performs the forward pass
+        :param num_segments: number of dendrite segments per out-unit
+        :param dim_context: length of the context vector;
+                            the same context will be applied to each segment
+        :param module_sparsity: sparsity applied over linear module;
+        :param dendrite_sparsity: sparsity applied transformation per unit per segment
+        :param dendrite_bias: whether or not dendrite activations have an additive bias
+        """
+        self.segments = None
+        super().__init__(module, sparsity=module_sparsity)
+
+        self.segments = DendriteSegments(
+            num_units=module.out_channels,
+            num_segments=num_segments,
+            dim_context=dim_context,
+            sparsity=dendrite_sparsity,
+            bias=dendrite_bias,
         )
+
+        self.rezero_weights()
+
+    def rezero_weights(self):
+        super().rezero_weights()
+        if self.segments is not None:  # only none at beggining of init
+            self.segments.rezero_weights()
+
+    def apply_dendrites(self, y, dendrite_activations):
+        """
```

Comment:
Shouldn't there be a forward() function that calls apply_dendrites? Then you can do `y = dendrite_layer(x, context)`.

Reply:
Yes, you're right. I just added `forward` (and copied from the 1d dendrite layers).